src/java/test/org/apache/zookeeper/server/NIOServerCnxnTest.java

```diff
@@ -68,4 +69,41 @@ public void testOperationsAfterCnxnClose() throws IOException,
             zk.close();
         }
     }
+
+    /**
+     * Mock extension of NIOServerCnxn to test for
+     * CancelledKeyException (ZOOKEEPER-2044).
+     */
+    private static class MockNIOServerCnxn extends NIOServerCnxn {
+        public MockNIOServerCnxn(NIOServerCnxn cnxn)
+                throws IOException {
+            super(cnxn.zkServer, cnxn.sock, cnxn.sk, cnxn.factory);
+        }
+
+        public void mockSendBuffer(ByteBuffer bb) throws Exception {
+            super.internalSendBuffer(bb);
+        }
+    }
+
+    @Test(timeout = 30000)
+    public void testValidSelectionKey() throws Exception {
+        int oldTimeout = ClientBase.CONNECTION_TIMEOUT;
+        ClientBase.CONNECTION_TIMEOUT = 3000;
+        final ZooKeeper zk = createClient();
```

Comment:
Thanks @hanm for the analysis and fixing it. Instead of directly changing the static value, how about simplifying the ZooKeeper client creation like below,

``final ZooKeeper zk = createZKClient(hostPort, 3000);``